File: src/services/aiChatService.ts
import { config } from '../config/env';
import type { RetrievedContext } from './messageService';
import { AppError } from '../utils/errors';

type ChatTurn = {
  role: 'user' | 'assistant' | 'system';
  content: string;
};

type ResponsesApiPayload = {
  model: string;
  input: Array<{
    role: 'user' | 'assistant' | 'system';
    content: Array<{ type: 'input_text'; text: string }>;
  }>;
};

type OpenAiResponsesApiResult = {
  output_text?: string;
  output?: Array<{
    content?: Array<{
      type?: string;
      text?: string;
    }>;
  }>;
  error?: {
    message?: string;
  };
};

function normalizeContext(context?: RetrievedContext): string {
  if (!context) {
    return '';
  }

  return `\n\nRetrieved context:\n${JSON.stringify(context)}`;
}

export async function generateAssistantReply(params: {
  history: ChatTurn[];
  userMessage: string;
  retrievedContext?: RetrievedContext;
}): Promise<string> {
  if (!config.openAiApiKey) {
    throw new AppError(500, 'OPENAI_API_KEY is not configured');
  }

  const userMessageWithContext = `${params.userMessage}${normalizeContext(params.retrievedContext)}`;

  const input: ResponsesApiPayload['input'] = [
    ...params.history.map((turn) => ({
      role: turn.role,
      content: [{ type: 'input_text' as const, text: turn.content }]
    })),
    {
      role: 'user',
      content: [{ type: 'input_text', text: userMessageWithContext }]
    }
  ];

  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), config.openAiTimeoutMs);

  let response: globalThis.Response;
  let data: OpenAiResponsesApiResult;

  try {
    response = await fetch(`${config.openAiBaseUrl}/responses`, {
      method: 'POST',
      headers: {
        Authorization: `Bearer ${config.openAiApiKey}`,
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({
        model: config.openAiModel,
        input
      } satisfies ResponsesApiPayload),
      signal: controller.signal
    });

    data = (await response.json()) as OpenAiResponsesApiResult;
  } catch (error) {
    if (error instanceof Error && error.name === 'AbortError') {
      throw new AppError(504, 'OpenAI request timed out');
    }
    throw new AppError(502, 'OpenAI request failed');
  } finally {
    clearTimeout(timeout);
  }

  if (!response.ok) {
    const message = data.error?.message || 'OpenAI request failed';
    throw new AppError(response.status, message);
  }

  const textFromOutput = data.output_text;
  const textFromContent = data.output
    ?.flatMap((item) => item.content || [])
    .find((contentItem) => contentItem.type === 'output_text' && typeof contentItem.text === 'string')?.text;

  const assistantText = textFromOutput || textFromContent;

  if (!assistantText) {
    throw new AppError(502, 'OpenAI response did not include assistant text');
  }

  return assistantText;
}

File: src/main/java/com/example/ragchat/service/AiChatService.java
package com.example.ragchat.service;

import com.example.ragchat.config.AppProperties;
import com.example.ragchat.entity.ChatMessage;
import com.example.ragchat.entity.SenderType;
import com.example.ragchat.exception.AppException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.stereotype.Service;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Service
public class AiChatService {
    private final AppProperties appProperties;
    private final ObjectMapper objectMapper;
    private final HttpClient httpClient;

    public AiChatService(AppProperties appProperties, ObjectMapper objectMapper) {
        this.appProperties = appProperties;
        this.objectMapper = objectMapper;
        this.httpClient = HttpClient.newBuilder().connectTimeout(Duration.ofSeconds(10)).build();
    }

    public String generateReply(List<ChatMessage> history, String message, Object retrievedContext) {
        if (appProperties.openAi().apiKey() == null || appProperties.openAi().apiKey().isBlank()) {
            throw new AppException(500, "OPENAI_API_KEY is not configured", "OPENAI_KEY_MISSING");
        }

        try {
            List<Map<String, Object>> input = new ArrayList<>();
            for (ChatMessage chatMessage : history.reversed()) {
                input.add(toInputItem(chatMessage.getSender(), chatMessage.getContent()));
            }

            String userContent = message;
            if (retrievedContext != null) {
                userContent = message + "\\n\\nRetrieved context:\\n" + objectMapper.writeValueAsString(retrievedContext);
            }
            input.add(toInputItem(SenderType.user, userContent));

            Map<String, Object> payload = new HashMap<>();
            payload.put("model", appProperties.openAi().model());
            payload.put("input", input);

            HttpRequest request = HttpRequest.newBuilder()
                    .uri(URI.create(appProperties.openAi().baseUrl() + "/responses"))
                    .timeout(appProperties.openAi().timeout())
                    .header("Authorization", "Bearer " + appProperties.openAi().apiKey())
                    .header("Content-Type", "application/json")
                    .POST(HttpRequest.BodyPublishers.ofString(objectMapper.writeValueAsString(payload)))
                    .build();

            HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
            JsonNode body = objectMapper.readTree(response.body());

            if (response.statusCode() >= 400) {
                String error = body.path("error").path("message").asText("OpenAI request failed");
                throw new AppException(response.statusCode(), error, "OPENAI_REQUEST_FAILED");
            }

            String outputText = body.path("output_text").asText();
            if (outputText != null && !outputText.isBlank()) {
                return outputText;
            }

            JsonNode output = body.path("output");
            for (JsonNode item : output) {
                for (JsonNode contentItem : item.path("content")) {
                    if ("output_text".equals(contentItem.path("type").asText())) {
                        String text = contentItem.path("text").asText();
                        if (!text.isBlank()) {
                            return text;
                        }
                    }
                }
            }

            throw new AppException(502, "OpenAI response did not include assistant text", "OPENAI_EMPTY_RESPONSE");
        } catch (AppException ex) {
            throw ex;
        } catch (Exception ex) {
            throw new AppException(502, "OpenAI request failed", "OPENAI_REQUEST_FAILED");
        }
    }

    private Map<String, Object> toInputItem(SenderType sender, String content) {
        Map<String, Object> inputItem = new HashMap<>();
        inputItem.put("role", sender.name());

        Map<String, Object> contentItem = new HashMap<>();
        contentItem.put("type", "input_text");
        contentItem.put("text", content);

        inputItem.put("content", List.of(contentItem));
        return inputItem;
    }
}
